### ISSUE TITLE
RBM autoencoding - no learning

### ISSUE THREAD

**Issue by szingerpeter** [deeplearning4j/deeplearning4j#2968]
Dear all,

I was about trying to create a Neural Network using RBM layers for auto encoding (in order to get familier with RBMs). However, I encountered an issue while trying to train the network.

For very different inputs, the model is giving either the **same outputs** or within a very short interval e.g: [0.34, 0.44].

I assume, it is a simple problem, but unfortunately was unable to find a solution to it, any help would be appreciated.

Here is the two models I was trying to train (both gave the same problem):
https://raw.githubusercontent.com/szingerpeter/DL4J./master/RbmClass_TwoLayer_RBM_RBM.java
https://raw.githubusercontent.com/szingerpeter/DL4J./master/RbmClass_TwoLayer_RBM_Output.java

Here is the dataset I am using:
https://raw.githubusercontent.com/szingerpeter/DL4J./master/RBM_x%3B2*x%3Bx%5E2.csv

Some settings out of the many I tried (regularization, normalization always true) for both models:
WeightInit: Xavier, Activation: Sigmoid, Opt.Algo: Stochastic_GD, Lossfunction: KL_Divergence
WeightInit: Relu_Uniform, Activation: Tanh, Opt.Algo: Line_GD, Lossfunction: MSE
WeightInit: Distribution, Activation: Identity, Opt.Algo: LBFGS, Lossfunction: Reconstruction_Crossentropy

Thank you for your time!

Best regards,
Peter

**Comment by szingerpeter** [comment]
Remark on (RBM + Output):
I am attaching the ![Screenshot of GUI of learning](https://cloud.githubusercontent.com/assets/19920281/23605459/18d7a630-025d-11e7-97d1-40e0572fc586.png)

Personally, it seems to be very cyclical to me even though the learningRate is set to 1e-9. (the graphs are cyclical with training of learningRate 1e-3 and 1 as well).

Moreover, I was trying to keep track of the training (with learningrate = 1), how the gradient, score, weights change over the training for each minibatch. Already at the beginning the gradient is relatively small (in the order of 1e-2) and decreases relatively continuously (converges to 0 for each parameter) even though the score does not tend to decrease over the training.

Moreover, after each iteration, the model parameters were compared to the parameters before the training, but did not change
('''if(wOld.toString().equalsIgnoreCase(model.paramTable().toString()))''' is in the for-each loop for the training).

Moreover, the model output and score looks very interesting as well after the training when I try to see how the model works (both the features and labels are normalized):
Original: [0.51, 0.51, 0.00]	 Rerepresentation: [109.79, -109.64, -89.42]	 Error: 7.972199249267578
Original: [0.61, 0.61, 0.05]	 Rerepresentation: [109.82, -109.66, -89.44]	 Error: 7.972199249267578
Original: [0.63, 0.63, 0.07]	 Rerepresentation: [109.82, -109.66, -89.44]	 Error: 7.972199249267578
Original: [0.16, 0.16, 0.48]	 Rerepresentation: [109.79, -109.64, -89.42]	 Error: 7.972199249267578
Original: [0.63, 0.63, 0.07]	 Rerepresentation: [109.82, -109.66, -89.44]	 Error: 7.972199249267578

**Comment by szingerpeter** [comment]
Remark on (RBM + RBM):
I am attaching the screenshot of the learning
Layer 0:
![GUI Screenshot learning layer 0](https://cloud.githubusercontent.com/assets/19920281/23609252/65ed8434-026d-11e7-9095-b042a72df151.png)
Layer 1:
![GUI Screenshot Layer 1](https://cloud.githubusercontent.com/assets/19920281/23609379/fdd9e4ea-026d-11e7-9219-2e4aa0d7c774.png)

Something weird happens here as well with learningRate = 1e-2 as you can see on the screenshot above. During the training, the score goes below 0 and stays negative. (The same thing happens with learningRate = 1)

Furthermore, during the training, the parameters do not change with comparison to the parameters before the training.
("if(wOld.toString().equalsIgnoreCase(model.paramTable().toString()))" is in the for-each loop for the training).

Moreover, the model gives this result:
Original: [0.32, 0.32, 0.14]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.43, 0.43, 0.02]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.84, 0.84, 0.47]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.65, 0.65, 0.09]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.08, 0.08, 0.70]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.90, 0.90, 0.63]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.87, 0.87, 0.56]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.50, 0.50, 0.00]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.66, 0.66, 0.11]	 Rerepresentation: 0.58	 Error: 0.0
Original: [0.53, 0.53, 0.00]	 Rerepresentation: 0.58	 Error: 0.0

**Comment by huitseeker** [comment]
Thanks for your report! It would be tremendously useful to have news about which versions of underlying software you are using. Would you mind confirming:
- project version (or branch if you are building off the repo),
- java version,
- CUDA version (if applicable),
- BLAS version (if applicable)
Thanks for your help !

**Comment by szingerpeter** [comment]
I'm using version 0.7.2 of both nd4j and deeplearning4j-core (I hope this answers your question)
JDK:
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-8u121-b13-0ubuntu1.16.04.2-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)

Thank you!

**Comment by agibsonccc** [comment]
Don't use kl deivergence on your output layer. You're combining relu weight init with tanh (hint: it's called relu for a reason ) and using random search algorithms. You've gone about trying the wrong combinations of parameters. There's no bug here. Stick to this guide: deeplearning4j.org/troubleshootingneuralnets The main parameters you should be playing with are learning rate and regularization here. For loss function/activation function/layer combinations stick to best practices in our examples.
